firefox_puppeteer/ui/windows.py

```diff
@@ -0,0 +1,486 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from time import sleep
+
+from marionette import By, Wait
+from marionette.errors import NoSuchWindowException
+from marionette.keys import Keys
+
+import firefox_puppeteer.errors as errors
+
+from ..api.l10n import L10n
+from ..base import BaseLib
+from ..decorators import use_class_as_property
+
+
+class Windows(BaseLib):
+
+    @property
+    def all(self):
+        """Retrieves a list of all open chrome windows.
+
+        :returns: List of :class:`BaseWindow`'s corresponding to the
+                  windows in `marionette.chrome_window_handles`.
+        """
+        return [self.create_window_instance(handle) for handle in
+                self.marionette.chrome_window_handles]
+
+    @property
+    def current(self):
+        """Retrieves the currently selected chrome window.
+
+        :returns: The :class:`BaseWindow` for the currently active window.
+        """
+        return self.create_window_instance(self.marionette.current_chrome_window_handle)
+
+    @property
+    def focused_chrome_window_handle(self):
+        """Returns the currently focused chrome window handle
+
+        :returns: The `window handle` of the focused chrome window.
+        """
+        with self.marionette.using_context('chrome'):
+            return self.marionette.execute_script("""
+              Cu.import("resource://gre/modules/Services.jsm");
+              var win = Services.wm.getMostRecentWindow("");
+              return win.QueryInterface(Ci.nsIInterfaceRequestor)
+                        .getInterface(Ci.nsIDOMWindowUtils)
+                        .outerWindowID.toString();
+            """)
+
+    def close(self, handle):
+        """Closes the chrome window with the given handle.
+
+        :param handle: The handle of the chrome window
+        """
+        self.switch_to(handle)
+
+        # TODO: Maybe needs to wait as handled via an observer
+        return self.marionette.close_chrome_window()
+
+    def close_all(self, exceptions=None):
+        """Closes all open chrome windows.
+
+        There is an optional `exceptions` list, which can be used to exclude
+        specific chrome windows from being closed.
+
+        :param exceptions: Optional, list or a single entry of handles or
+         :class:`BaseWindow` instances not to close
+        """
+        handles_to_keep = exceptions or []
+        if not isinstance(handles_to_keep, list):
+            handles_to_keep = [handles_to_keep]
```

Comment:
Sigh. I really don't like APIs that let you have multiple input types, but I know from prior debates that I'm probably in the minority. I'll leave it to your discretion.